cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1674 @@
+}
+
+// Main flows
+
+// Update lossless pg on a port after an PG has been installed on the port
+// Called when pg updated from CONFIG_DB
+// Key format: BUFFER_PG:<port>:<pg>
+task_process_status BufferMgrDynamic::doUpdatePgTask(const string &pg_key, const string &port)
+{
+    string value;
+    port_info_t &portInfo = m_portInfoLookup[port];
+    auto &bufferPg = m_portPgLookup[port][pg_key];
+    task_process_status task_status = task_process_status::task_success;
+
+    switch (portInfo.state)
+    {
+    case PORT_READY:
+        // Not having profile_name but both speed and cable length have been configured for that port
+        // This is because the first PG on that port is configured after speed, cable length configured
+        // Just regenerate the profile
+        task_status = refreshPriorityGroupsForPort(port, portInfo.speed, portInfo.cable_length, portInfo.mtu, pg_key);
+        if (task_status != task_process_status::task_success)
+            return task_status;
+
+        break;
+
+    case PORT_INITIALIZING:
+        if (bufferPg.dynamic_calculated)
+        {
+            SWSS_LOG_NOTICE("Skip setting BUFFER_PG for %s because port's info isn't ready for dynamic calculation", pg_key.c_str());
+        }
+        else
+        {
+            task_status = refreshPriorityGroupsForPort(port, portInfo.speed, portInfo.cable_length, portInfo.mtu, pg_key);
+            if (task_status != task_process_status::task_success)
+                return task_status;
+        }
+        break;
+
+    default:
+        // speed and cable length hasn't been configured
+        // In that case, we just skip the this update and return success.
+        // It will be handled after speed and cable length configured.
+        SWSS_LOG_NOTICE("Skip setting BUFFER_PG for %s because port's info isn't ready for dynamic calculation", pg_key.c_str());
+        return task_process_status::task_success;
+    }
+
+    if (bufferPg.static_configured && bufferPg.dynamic_calculated)
+    {
+        auto &profile = m_bufferProfileLookup[bufferPg.configured_profile_name];
+        profile.port_pgs.insert(pg_key);
+    }
+
+    return task_process_status::task_success;
+}
+
+//Remove the currently configured lossless pg
+task_process_status BufferMgrDynamic::doRemovePgTask(const string &pg_key, const string &port)
+{
+    auto &bufferPgs = m_portPgLookup[port];
+    buffer_pg_t &bufferPg = bufferPgs[pg_key];
+    port_info_t &portInfo = m_portInfoLookup[port];
+
+    // Remove the PG from APPL_DB
+    string null_str("");
+    updateBufferPgToDb(pg_key, null_str, false);
+
+    SWSS_LOG_NOTICE("Remove BUFFER_PG %s (profile %s, %s)", pg_key.c_str(), bufferPg.running_profile_name.c_str(), bufferPg.configured_profile_name.c_str());
+
+    // recalculate pool size
+    checkSharedBufferPoolSize();
+
+    if (!portInfo.speed.empty() && !portInfo.cable_length.empty())
+        portInfo.state = PORT_READY;
+    else
+        portInfo.state = PORT_INITIALIZING;
```

Comment:
In this case, when does the port goes to port_Ready state after lossless PG is removed?

Reply:
For headroom override (statically configured profile on a port), it will go to `PORT_READY` once both `speed` and `cable length` are configured.
For dynamic calculated profile, this sentence won't be reached because only if both `speed` and `cable length` are provided will the headroom (profile) be calculated.